### ISSUE TITLE
【论文复现营】请问为什么要使用类似这样的操作：tw[torch_key].detach().numpy() 

### ISSUE THREAD

**Issue by skywalk163** [PaddlePaddle/X2Paddle#610]
# 问题
听课的时候，老师讲到pytorch权重转换到paddlepaddle的时候，当时讲的是要转置一下，代码是tw[torch_key].detach().numpy() ，但是我看api讲detach是copy子图啊，不理解这里的意思，
希望能解释一下。

在x2paddle源代码找到，在x2paddle/project_convertor/pytorch/convert.py文件的convert_params代码块也是用了detach()操作：

···
def convert_params(params_path):
    import torch
    import paddle
    params = torch.load(params_path, map_location=torch.device('cpu'))
    new_params = dict()
    bn_w_name_list = list()
    for k, v in params.items():
        if k.endswith(".running_mean"):
            new_params[k.replace(".running_mean", "._mean")] = v.detach().numpy(
            )
        elif k.endswith(".running_var"):
            new_params[k.replace(".running_var", "._variance")] = v.detach(
            ).numpy()
            bn_w_name_list.append(k.replace(".running_var", ".weight"))
        else:
            new_params[k] = v.detach().numpy()
    for k, v in new_params.items():
        if len(v.shape) == 2 and k.endswith(
                ".weight") and k not in bn_w_name_list:
            new_params[k] = v.T
    paddle.save(new_params,
                params_path.replace(".pth", ".pdiparams").replace(
                    ".pt", ".pdiparams").replace(".ckpt", ".pdiparams"))
’‘’

**Comment by SunAhong1993** [comment]
这里是因为PyTorch保存参数的类型为torch.Tensor，而Paddle则是numpy格式保存，因此需要将子图复制出来，转换成numpy格式的参数进行保存。

**Comment by skywalk163** [comment]
谢谢，那这句就是找到fc层并转置吧？ 飞桨和torch的fc层格式正好是转置关系？     
for k, v in new_params.items():
        if len(v.shape) == 2 and k.endswith(
                ".weight") and k not in bn_w_name_list:
            new_params[k] = v.T

**Comment by SunAhong1993** [comment]
> 谢谢，那这句就是找到fc层并转置吧？ 飞桨和torch的fc层格式正好是转置关系？
> for k, v in new_params.items():
> if len(v.shape) == 2 and k.endswith(
> ".weight") and k not in bn_w_name_list:
> new_params[k] = v.T

是的，没错的
